internal/internal_workflow.go

```diff
@@ -517,24 +518,17 @@ func (d *syncWorkflowDefinition) Execute(env WorkflowEnvironment, header *common
 	})
 
 	getWorkflowEnvironment(d.rootCtx).RegisterSignalHandler(func(name string, input *commonpb.Payloads) error {
-		eo := getWorkflowEnvOptions(d.rootCtx)
-		// Notify interceptor, note that because channel send operation below is non-blocking, we can not pass a full closure
-		// that would encapsulate signal processing. Hence this call is just a "forked" notification that signal has been received.
-		err := envInterceptor.inboundInterceptor.ProcessSignal(d.rootCtx, name, input)
-		if err != nil {
-			return err
-		}
-		// We don't want this code to be blocked ever, using sendAsync().
-		ch := eo.getSignalChannel(d.rootCtx, name).(*channelImpl)
-		ok := ch.SendAsync(input)
-		if !ok {
-			panic(fmt.Sprintf("Exceeded channel buffer size for signal: %v", name))
-		}
-		return nil
+		// TODO(cretz): We decode ahead of time for HandleQuery but we don't here. Is
+		// this inconsistency acceptable? Is it ok to pretend that "Arg" is
+		// interface{} when we know it will always be *commonpb.Payloads?
+		// See https://github.com/temporalio/proposals/pull/45/files#r732871182
```

Comment:
I would like some feedback/discussion here. We cannot decode ahead of time for signals, but like the old interceptors, we leave the `Arg` as `interface{}` even though we know it's always a payload. Shall we change its type since we know it's always a payload and we can't decode it?

Reply:
I am now thinking the parameter is best as a `converter.EncodedValue`.

Reply:
:exploding_head:

Reply:
Changed `Arg` type to a raw payload

Reply:
I think you can get rid of the TODO now?